cpp/src/arrow/dataset/file_parquet.cc

```diff
@@ -248,7 +283,12 @@ Result<std::vector<int>> InferColumnProjection(const parquet::arrow::FileReader&
   }
 
   std::vector<int> columns_selection;
-  for (const auto& ref : field_refs) {
+  for (auto& ref : field_refs) {
+    // In the (unlikely) absence of a known dataset schema, we require that all
+    // materialized refs are named.
+    if (options.dataset_schema) {
```

Comment:
Heh, at this point, if we don't have a dataset schema, I think something has gone very wrong.

Reply:
Haha I figured as much... That being said, some of the existing tests do leave it null.